modules/graphql_core/src/Plugin/Deriver/EntityFieldDeriver.php

```diff
@@ -8,43 +8,49 @@
 use Drupal\graphql_core\Plugin\GraphQL\Fields\EntityField;
 use Drupal\graphql_core\Plugin\GraphQL\Types\EntityFieldType;
 
-// TODO Reduce single-property fields to scalars.
+// TODO Write tests for entity reference graph traversal.
+
+// TODO Should we expose config entities?
 
 // TODO Convert timestamps to strings?
 
 /**
  * Deriver for RawValue fields.
  */
-class EntityFieldDeriver extends EntityFieldDeriverBase {
+class EntityFieldDeriver extends EntityFieldDeriverWithTypeMapping {
 
   /**
    * {@inheritdoc}
    */
-  protected function getBaseFieldDefinition($entityTypeId, BaseFieldDefinition $baseFieldDefinition, array $basePluginDefinition) {
-    $fieldName = $baseFieldDefinition->getName();
+  protected function getDerivativesFromPropertyDefinitions($entityTypeId, FieldStorageDefinitionInterface $definition, array $basePluginDefinition, $bundleId = NULL) {
+    $fieldName = $definition->getName();
 
-    $this->derivatives["$entityTypeId-$fieldName"] = [
+    $derivative = [
       'types' => [StringHelper::camelCase([$entityTypeId])],
       'name' => EntityField::getId($fieldName),
-      'multi' => $baseFieldDefinition->isMultiple(),
-      'field' => $fieldName,
-      'type' => EntityFieldType::getId($entityTypeId, $fieldName),
-    ] + $basePluginDefinition;
-  }
-
-  /**
-   * {@inheritdoc}
-   */
-  protected function getConfigFieldDefinition($entityTypeId, $bundleId, FieldStorageDefinitionInterface $storage, array $basePluginDefinition) {
-    $fieldName = $storage->getName();
-
-    $this->derivatives["$entityTypeId-$bundleId-$fieldName"] = [
-      'types' => [StringHelper::camelCase([$entityTypeId, $bundleId])],
-      'name' => EntityField::getId($fieldName),
-      'multi' => $storage ? $storage->getCardinality() != 1 : FALSE,
+      'multi' => $definition->isMultiple(),
       'field' => $fieldName,
-      'type' => EntityFieldType::getId($entityTypeId, $fieldName),
-    ] + $basePluginDefinition;
+    ];
+
+    $properties = $definition->getPropertyDefinitions();
+    if (count($properties) == 1) {
+      // Flatten the structure for single-property fields.
+      /** @var \Drupal\Core\TypedData\DataDefinitionInterface $property */
+      $property = reset($properties);
+      $keys = array_keys($properties);
+
+      $derivative['type'] = $this->typeMapper->typedDataToGraphQLFieldType($property);
+      $derivative['property'] = reset($keys);
+    }else {
+      $derivative['type'] = EntityFieldType::getId($entityTypeId, $fieldName);
+    }
+
+    if ($bundleId) {
```

Comment:
Always check empty / !empty instead of just checking the truthy value of a parameter.